src/core/internal/array/equality.d

```diff
@@ -10,125 +10,48 @@
 
 module core.internal.array.equality;
 
-// compiler frontend lowers dynamic array comparison to this
-bool __ArrayEq(T1, T2)(T1[] a, T2[] b)
+// The compiler lowers `lhs == rhs` to `__equals(lhs, rhs)` for
+// * dynamic arrays,
+// * (most) arrays of different (unqualified) element types, and
+// * arrays of structs with custom opEquals.
+bool __equals(T1, T2)(scope T1[] lhs, scope T2[] rhs)
 {
-    if (a.length != b.length)
-        return false;
-    foreach (size_t i; 0 .. a.length)
-    {
-        if (a[i] != b[i])
-            return false;
-    }
-    return true;
-}
-
- // `lhs == rhs` lowers to `__equals(lhs, rhs)` for dynamic arrays
-bool __equals(T1, T2)(T1[] lhs, T2[] rhs)
-{
-    import core.internal.traits : Unqual;
-    alias U1 = Unqual!T1;
-    alias U2 = Unqual!T2;
-
-    static @trusted ref R at(R)(R[] r, size_t i) { return r.ptr[i]; }
-    static @trusted R trustedCast(R, S)(S[] r) { return cast(R) r; }
-
     if (lhs.length != rhs.length)
         return false;
 
-    if (lhs.length == 0 && rhs.length == 0)
+    if (lhs.length == 0)
         return true;
 
-    static if (is(U1 == void) && is(U2 == void))
-    {
-        return __equals(trustedCast!(ubyte[])(lhs), trustedCast!(ubyte[])(rhs));
-    }
-    else static if (is(U1 == void))
-    {
-        return __equals(trustedCast!(ubyte[])(lhs), rhs);
-    }
-    else static if (is(U2 == void))
-    {
-        return __equals(lhs, trustedCast!(ubyte[])(rhs));
-    }
-    else static if (!is(U1 == U2))
-    {
-        // This should replace src/object.d _ArrayEq which
-        // compares arrays of different types such as long & int,
-        // char & wchar.
-        // Compiler lowers to __ArrayEq in dmd/src/opover.d
-        foreach (const u; 0 .. lhs.length)
-        {
-            if (at(lhs, u) != at(rhs, u))
-                return false;
-        }
-        return true;
-    }
-    else static if (__traits(isIntegral, U1))
+    // Use memcmp if applicable to both element types and the size matches.
+    static if (T1.sizeof == T2.sizeof && useMemcmp!T1 && useMemcmp!T2)
```

Comment:
Need to also check `__traits(isUnsigned, T1) == __traits(isUnsigned, T2)` or else:

```d
ubyte[] a = [cast(ubyte) 0xff];
byte[] b = [cast(byte) 0xff];
assert(a != b); // Currently passes but will fail after this PR.
```

Reply:
Right, thx. That only affects integers < 32 bit though, due to int promotion.